WordPress/gutenberg
# Make text/number inputs of `ColorPicker` keep draft values while focused

## What?
In support of the changes to `InputControl` in #40518 this ensures the text and number inputs of `ColorPicker` will keep a "drafted" value while the input is focused (like they do in trunk but without relying on the current `InputControl` behavior).

## Why?
To avoid interference with typing a value in the input and preserve instant updates during entry.

## How?
Adds logic (in a hook) to prefer a draft value over the state while focused yet restore control to props after if a change comes in that wasn't initiated from the input (e.g. global undo).

## Testing Instructions
1. Open a Post or Page.
2. Insert a Heading Block.
3. Open the Text color settings for the block.
4. Open the Custom Color Picker.
5. Show the detailed inputs.
6. Clear the input.
7. Enter a color value.
8. Verify that the value entered is kept as is and the visible color changes as different values are entered.
9. Enter a shorthand value with a letter e.g. `c03`.
10. Click outside the input.
11. Verify that the value becomes the longhand form and its letters are changed to uppercase.
12. Focus the input.
13. Press the keyboard shortcut for undo.
14. Verify that the actual selected color in the parent dialog reverts to what it was before any value changes*

*The result of the last step can be confusing because the color picker itself will revert to its last local state and instead of the editor’s revision as can be seen in its parent dialog . This is an existing issue with the component and an edge case besides.

## Screenshots or screencast

### Issue with the hex input of `ColorPicker` from #40518
https://user-images.githubusercontent.com/9000376/169936732-ac9304f3-f73e-41bf-bbe4-cd1467e3cae1.mp4

### The hex input on this branch
https://user-images.githubusercontent.com/9000376/169936875-14c8afe8-2766-41d0-8679-7f18197ea761.mp4

While I recorded a different set of actions they both include typing into the hex input so this demonstrates the fix. It also demonstrates the undo works (aside from the caveat mentioned in testing instructions) even while the input is focused.

### Issue with `InputWithSlider` of `ColorPicker` from #40518 
https://user-images.githubusercontent.com/9000376/170106420-141339bd-d1f6-41ca-9ecc-1fc7943f213d.mp4

This is the same issue with the hex input but isn't likely to be as commonly encountered. What is also seen in the screen recording is a very related but separate issue with these controls #36703.

### `InputWithSlider` on this branch
https://user-images.githubusercontent.com/9000376/170107232-5f808854-82ae-4305-866b-9ed1d41ac20f.mp4
